kythe/go/serving/xrefs/columnar/columnar_encoding.go

```diff
@@ -0,0 +1,366 @@
+	key, err := keys.Append(prefix, file, columnarDecorationsTextGroup, t.StartOffset, t.EndOffset)
+	if err != nil {
+		return nil, err
+	}
+	val, err := proto.Marshal(&xspb.FileDecorations_Text{Text: t.Text})
+	if err != nil {
+		return nil, err
+	}
+	return &KV{key, val}, nil
+}
+
+func encodeDecorTarget(prefix []byte, file *spb.VName, t *xspb.FileDecorations_Target) (*KV, error) {
+	key, err := keys.Append(prefix, file,
+		columnarDecorationsTargetGroup, t.StartOffset, t.EndOffset,
+		t.GetGenericKind(), int32(t.GetKytheKind()), t.Target)
+	if err != nil {
+		return nil, err
+	}
+
+	val, err := proto.Marshal(&xspb.FileDecorations_Target{})
+	if err != nil {
+		return nil, err
+	}
+	return &KV{key, val}, nil
+}
+
+func encodeDecorTargetOverride(prefix []byte, file *spb.VName, to *xspb.FileDecorations_TargetOverride) (*KV, error) {
+	key, err := keys.Append(prefix, file,
+		columnarDecorationsTargetOverrideGroup, to.Overridden, int32(to.Kind), to.Overriding)
+	if err != nil {
+		return nil, err
+	}
+	val, err := proto.Marshal(&xspb.FileDecorations_TargetOverride{})
+	if err != nil {
+		return nil, err
+	}
+	return &KV{key, val}, nil
+}
+func encodeDecorTargetNode(prefix []byte, file *spb.VName, tn *xspb.FileDecorations_TargetNode) (*KV, error) {
+	key, err := keys.Append(prefix, file, columnarDecorationsTargetNodeGroup, tn.Node.Source)
+	if err != nil {
+		return nil, err
+	}
+	val, err := proto.Marshal(tn)
+	if err != nil {
+		return nil, err
+	}
+	return &KV{key, val}, nil
+}
+func encodeDecorTargetDefinition(prefix []byte, file *spb.VName, td *xspb.FileDecorations_TargetDefinition) (*KV, error) {
+	key, err := keys.Append(prefix, file, columnarDecorationsTargetDefinitionGroup, td.Target)
+	if err != nil {
+		return nil, err
+	}
+	val, err := proto.Marshal(&xspb.FileDecorations_TargetDefinition{Definition: td.Definition})
+	if err != nil {
+		return nil, err
+	}
+	return &KV{key, val}, nil
+}
+func encodeDecorDefinitionLocation(prefix []byte, file *spb.VName, def *xspb.FileDecorations_DefinitionLocation) (*KV, error) {
+	// TODO(schroederc): use VNames throughout pipeline
+	defVName, err := kytheuri.ToVName(def.Location.Ticket)
+	if err != nil {
+		return nil, err
+	}
+	key, err := keys.Append(prefix, file, columnarDecorationsDefinitionLocationGroup, defVName)
+	if err != nil {
+		return nil, err
+	}
+	val, err := proto.Marshal(def)
+	if err != nil {
+		return nil, err
+	}
+	return &KV{key, val}, nil
+}
+func encodeDecorOverride(prefix []byte, file *spb.VName, o *xspb.FileDecorations_Override) (*KV, error) {
+	key, err := keys.Append(prefix, file, columnarDecorationsOverrideGroup, o.Override)
+	if err != nil {
+		return nil, err
+	}
+	val, err := proto.Marshal(&xspb.FileDecorations_Override{MarkedSource: o.MarkedSource})
+	if err != nil {
+		return nil, err
+	}
+	return &KV{key, val}, nil
+}
+func encodeDecorDiagnostic(prefix []byte, file *spb.VName, d *xspb.FileDecorations_Diagnostic) (*KV, error) {
+	val, err := proto.Marshal(d)
```

Comment:
Can you move this down a few lines to match the pattern of the other encode* functions?

Reply:
The value is marshaled first because the key needs its hash.